In [12]:
import pandas as pd
from sqlalchemy import create_engine

In [13]:
pd.set_option('display.max_colwidth', 100)
pd.set_option('display.float_format', '{:,.2f}'.format)

In [14]:
engine = create_engine("postgresql://XXXX_user:PPPPPPPPP@xxx.xxx.196.221/dddd_db")

1. Посчитай суммарную рекламную выручку и суммарные затраты для каждого бренда.

In [15]:
sql = """
select
    brand,
    sum (cost_nds) sum_cost,
    sum(revenue) sum_revenue
from ad_stats adst
join products pr
on adst.sku=pr.sku
group by brand
"""

pd.read_sql(sql, engine)

,brand,sum_cost,sum_revenue
0,Synergetic,"45,699,641.06","459,497,344.03"
1,SYNERGETIC,"102,324.01","805,311.09"
2,4fresh,"180,538.18","333,939.63"
3,OXYDENT,"205,305.07","651,252.26"
4,SMART FOX,"1,768,426.30","12,608,106.69"


Видим, что один бренд записан двумя разными регистрами. Приведем к единому стилю.

In [16]:
sql = """
select
    brand,
    sum (cost_nds) sum_cost,
    sum(revenue) sum_revenue
from ad_stats adst
join (
     select
         sku,
         upper(brand) brand from products
     ) pr
on adst.sku=pr.sku
group by brand
order by sum_revenue desc
"""

pd.read_sql(sql, engine)

,brand,sum_cost,sum_revenue
0,SYNERGETIC,"45,801,965.07","460,302,655.12"
1,SMART FOX,"1,768,426.30","12,608,106.69"
2,OXYDENT,"205,305.07","651,252.26"
3,4FRESH,"180,538.18","333,939.63"


2. Как ты думаешь, каким образом можно определить эффективность рекламы?

Как правило (с точки зрения финансов), это ROI - возврат на инвестицию.

3. Используя метрику из пункта 2, напиши запрос, который выведет Топ-10 категорий товаров с самой эффективной рекламой.

In [17]:
# Логично считать по прибыли (т.к. в товарной выручке есть себестоимость)
# но есть данные только по выручке, поэтому посчитаем по ней

# В некоторых категориях сумма затрат равна нулю, поэтому сделал фильтр

sql = """
select
    final_category,
    ( ( sum(sum_revenue) - sum(sum_cost) ) / sum(sum_cost) * 100 ) roi
from (
          select
                final_category,
                sum (cost_nds) sum_cost,
                sum(revenue) sum_revenue
          from ad_stats adst
          join products pr
          on adst.sku=pr.sku
          group by final_category
          having sum (cost_nds) > 0
) t
group by final_category
order by roi desc
limit 10
"""

pd.read_sql(sql, engine)

,final_category,roi
0,ПММ,"2,108.96"
1,Пол,"1,584.10"
2,Кусковое мыло,"1,582.22"
3,Очищение для лица,"1,572.08"
4,Шампуни,"1,412.43"
5,Сантехника,"1,409.93"
6,Освежители воздуха,"1,392.31"
7,Порошки,"1,137.96"
8,Гели для стирки,"1,106.13"
9,Маски для волос,"1,074.02"


4. Для каждого товара в категории Гели для стирки посчитай конверсии на каждом этапе воронки (суммируя данные за весь период). Выдели проблемные места, предложи варианты по улучшению показателей.

a. TR (Click-through rate) = ​​​​clicks / impressions​​​​ * 100%

In [18]:
sql = """
select
    pr.sku,
    pr.name,
    pr.brand,
    sum(clicks) / sum(impressions) * 100 tr
from ad_stats adst
join products pr
on adst.sku=pr.sku
where pr.final_category = 'Гели для стирки'
group by pr.sku
order by tr desc
"""

pd.read_sql(sql, engine)

,sku,name,brand,tr
0,539878217,"Гель для стирки белья универсальный SYNERGETIC 5 л 165 стирок, жидкий порошок, порошок стиральны...",Synergetic,2.89
1,1380046468,"Гель для стирки SYNERGETIC 3 IN 1 ""Магическая орхидея"", 5 л, жидкий порошок, порошок стиральный ...",Synergetic,2.78
2,3122752912,Гель для стирки FRESH&CLEAN 5 л (AROMA EMOTIONS Табак-Ваниль),Synergetic,2.69
3,540376152,"Гель для стирки SYNERGETIC 2в1 c пятновыводителем 5л, 165 стирок, жидкий порошок, порошок стирал...",Synergetic,2.49
4,2189036231,"Гель для стирки цветного белья SYNERGETIC COLOR, 5 л (166+ стирок)",Synergetic,1.91
5,2854972250,Гель для стирки 3в1 Цветочная ваниль 5 л,SMART FOX,1.50
6,2067258541,"Гель для стирки SMART FOX/SYNERGETIC UNIVERSAL, концентрат, 5 л, жидкий порошок, порошок стираль...",SMART FOX,1.27
7,2067262831,"Гель для стирки SMART FOX/SYNERGETIC COLOR Горный эдельвейс, концентрат, 5 л, жидкий порошок, по...",SMART FOX,0.94
8,2067213439,"Гель для стирки SMART FOX/SYNERGETIC COLOR Весеннее утро, концентрат, 5 л, жидкий порошок, порош...",SMART FOX,0.76


Видно, что проседает кликабельность у SMART FOX. Скорее всего, это из-за того, что он не такой узнаваемый как Synergetic. Нужно каким-то образом повышать узнаваемость.

b. CR из карточки в корзину = ​​​​to_cart / clicks​​​​ * 100%

In [19]:
sql = """
select
    pr.sku,
    pr.name,
    pr.brand,
    sum(to_cart) / sum(clicks) * 100 tr
from ad_stats adst
join products pr
on adst.sku=pr.sku
where pr.final_category = 'Гели для стирки'
group by pr.sku
order by tr desc
"""

pd.read_sql(sql, engine)

,sku,name,brand,tr
0,1380046468,"Гель для стирки SYNERGETIC 3 IN 1 ""Магическая орхидея"", 5 л, жидкий порошок, порошок стиральный ...",Synergetic,34.94
1,540376152,"Гель для стирки SYNERGETIC 2в1 c пятновыводителем 5л, 165 стирок, жидкий порошок, порошок стирал...",Synergetic,33.30
2,539878217,"Гель для стирки белья универсальный SYNERGETIC 5 л 165 стирок, жидкий порошок, порошок стиральны...",Synergetic,30.92
3,2067262831,"Гель для стирки SMART FOX/SYNERGETIC COLOR Горный эдельвейс, концентрат, 5 л, жидкий порошок, по...",SMART FOX,27.52
4,2067258541,"Гель для стирки SMART FOX/SYNERGETIC UNIVERSAL, концентрат, 5 л, жидкий порошок, порошок стираль...",SMART FOX,26.99
5,2189036231,"Гель для стирки цветного белья SYNERGETIC COLOR, 5 л (166+ стирок)",Synergetic,26.46
6,2854972250,Гель для стирки 3в1 Цветочная ваниль 5 л,SMART FOX,22.51
7,2067213439,"Гель для стирки SMART FOX/SYNERGETIC COLOR Весеннее утро, концентрат, 5 л, жидкий порошок, порош...",SMART FOX,17.93
8,3122752912,Гель для стирки FRESH&CLEAN 5 л (AROMA EMOTIONS Табак-Ваниль),Synergetic,17.82


Два нижних заметно остатют. Так сходу сложно сказать, почему эти товары хуже зашли с первого взляда. Возможно, было не совсем подходящее описание, у клиента не оправдались ожидания, или какие-то технические затыки в карточке товара.

c. CR из корзины в заказ= ​​​​orders / to_cart​​​​ * 100%

In [20]:
sql = """
select
    pr.sku,
    pr.name,
    pr.brand,
    sum(orders) / sum(to_cart) * 100 tr
from ad_stats adst
join products pr
on adst.sku=pr.sku
where pr.final_category = 'Гели для стирки'
group by pr.sku
order by tr desc
"""

pd.read_sql(sql, engine)

,sku,name,brand,tr
0,539878217,"Гель для стирки белья универсальный SYNERGETIC 5 л 165 стирок, жидкий порошок, порошок стиральны...",Synergetic,52.02
1,540376152,"Гель для стирки SYNERGETIC 2в1 c пятновыводителем 5л, 165 стирок, жидкий порошок, порошок стирал...",Synergetic,46.22
2,1380046468,"Гель для стирки SYNERGETIC 3 IN 1 ""Магическая орхидея"", 5 л, жидкий порошок, порошок стиральный ...",Synergetic,42.71
3,2067258541,"Гель для стирки SMART FOX/SYNERGETIC UNIVERSAL, концентрат, 5 л, жидкий порошок, порошок стираль...",SMART FOX,35.18
4,3122752912,Гель для стирки FRESH&CLEAN 5 л (AROMA EMOTIONS Табак-Ваниль),Synergetic,33.51
5,2189036231,"Гель для стирки цветного белья SYNERGETIC COLOR, 5 л (166+ стирок)",Synergetic,33.33
6,2067262831,"Гель для стирки SMART FOX/SYNERGETIC COLOR Горный эдельвейс, концентрат, 5 л, жидкий порошок, по...",SMART FOX,31.11
7,2854972250,Гель для стирки 3в1 Цветочная ваниль 5 л,SMART FOX,24.65
8,2067213439,"Гель для стирки SMART FOX/SYNERGETIC COLOR Весеннее утро, концентрат, 5 л, жидкий порошок, порош...",SMART FOX,24.35


У sku 2067213439 все показатели находятся внизу списков. Что-то с ним не так, нужно изучать.

5. Посчитай ежедневную выручку для бренда Synergetic, выручку за предыдущий день и прирост день ко дню. В результирующей таблице должны быть 4 колонки:  
a. Дата.  
b. Выручка за этот день.  
c. Выручка за **предыдущий день** (в соседней колонке).  
d. Процентный прирост (Growth %) день к дню.



In [21]:
sql = """
with
    products_upper as (
        select
            sku,
            upper(brand) brand from products
        where upper (brand)='SYNERGETIC'
),
    revenue_by_day as (
        select
            day,
            sum(revenue) revenue
        from ad_stats adst
        join products_upper
        on adst.sku=products_upper.sku
        group by day
)
select
    day,
    revenue,
    LAG(revenue, 1, null) OVER (ORDER BY day) previous,
    (revenue / (LAG(revenue, 1, null) OVER (ORDER BY day)) -1 ) * 100 as growth
from revenue_by_day
"""

pd.read_sql(sql, engine)

,day,revenue,previous,growth
0,2026-01-01,"16,607,352.71",NaN,NaN
1,2026-01-02,"27,593,355.76","16,607,352.71",66.15
2,2026-01-03,"27,084,645.25","27,593,355.76",-1.84
3,2026-01-04,"28,608,884.65","27,084,645.25",5.63
4,2026-01-05,"30,574,593.72","28,608,884.65",6.87
5,2026-01-06,"31,362,053.99","30,574,593.72",2.58
6,2026-01-07,"29,353,439.66","31,362,053.99",-6.40
7,2026-01-08,"30,897,625.27","29,353,439.66",5.26
8,2026-01-09,"30,688,239.67","30,897,625.27",-0.68
9,2026-01-10,"32,824,968.94","30,688,239.67",6.96


In [22]:
sql = """
with
    revenue_by_day as (
        select
            day,
            sum(revenue) revenue
        from ad_stats adst
        join products
        on adst.sku=products.sku
        where brand='SYNERGETIC' or brand='Synergetic'
        group by day
)
select
    day,
    revenue,
    LAG(revenue, 1, null) OVER (ORDER BY day) previous,
    (revenue / (LAG(revenue, 1, null) OVER (ORDER BY day)) -1 ) * 100 as growth
from revenue_by_day
"""

pd.read_sql(sql, engine)

,day,revenue,previous,growth
0,2026-01-01,"16,607,352.71",NaN,NaN
1,2026-01-02,"27,593,355.76","16,607,352.71",66.15
2,2026-01-03,"27,084,645.25","27,593,355.76",-1.84
3,2026-01-04,"28,608,884.65","27,084,645.25",5.63
4,2026-01-05,"30,574,593.72","28,608,884.65",6.87
5,2026-01-06,"31,362,053.99","30,574,593.72",2.58
6,2026-01-07,"29,353,439.66","31,362,053.99",-6.40
7,2026-01-08,"30,897,625.27","29,353,439.66",5.26
8,2026-01-09,"30,688,239.67","30,897,625.27",-0.68
9,2026-01-10,"32,824,968.94","30,688,239.67",6.96
